ggml-quants.c

```diff
@@ -3306,6 +3324,53 @@ size_t quantize_q8_0(const float * restrict src, void * restrict dst, int64_t nr
     return nrow * row_size;
 }
 
+size_t quantize_i2_s(const float * restrict src, void * restrict dst, int64_t nrow, int64_t n_per_row, const float * quant_weights) {
+    // 2 bits per weight
+    UNUSED(quant_weights);
+
+    size_t row_size = ggml_row_size(GGML_TYPE_I2_S, n_per_row);
+
+    int n = nrow * n_per_row;
+
+    // f32 -> q8
+    double i2_scale = 0;
+    for (int i=0; i<n; i++) {
+        if (fabs((double)(src[i])) > 1e-6) {
+            i2_scale = (double)src[i];
+        }
+    }
+
+    uint8_t* q8 = (uint8_t*)dst;
+    for (int i=0; i<n; i++) {
+        if (fabs((double)(src[i])) < 1e-6) {
+            q8[i] = 0;
+            continue;
+        }
+        q8[i] = (double)src[i] * i2_scale > 0 ? 1 : 3;
+    }
+
+    // q8 -> 0, 1, 3
+    //       |  |  |
+    //       0, 1,-1
+
+    uint8_t* i2_weight = (uint8_t*)dst;
+    for (int i=0; i<n; i++) {
+        int group_idx = i / 4;
+        int group_pos = i % 4;
+        uint8_t temp = (q8[i] << (6 - 2 * group_pos));
+        q8[i] = 0;
+        i2_weight[group_idx] |= temp;
+    }
+
+    float* scale_ptr = (float*)((char*)i2_weight + n / 4);
+    for (int i=0; i<8; i++) {
+        scale_ptr[i] = i2_scale;
+    }
+
+    // 32B for scale
+    return nrow * row_size / 4 + 32;
```

Comment:
Well, personally, I still think the tensor-wide scale is a better way for bitnet. At least for the 2bit compaction, 2.25 bpw means a around 10% model size waste, and it's kind of beyond what is acceptable.

Reply:
Since 2 bits is already wasting 20% of the tensor size compared to the 1.6 bpw ideal for ternary, maybe there could be a way to still make this a block-wise quant (e.g. 4 elements per block of 1 byte), and have a row-wise/tensor-wise scale by somehow encoding it in unused bits in the first few blocks of each row? Might be a bad idea though, but I don't know yet why. (This would require asserting a minimal row size in the quantize function.)

Because 4 ternary values fit in 7 bits (3^4 == 81 < 128), and you're already using a lookup table to expand the packed bits into 4 bytes, this could let the SIMD vec_dot stay pretty much identical to how it is now, except maybe it could include the scaling in its result?

Not sure yet how to pack the scale in `i8_s`, though, or some other way to let `ggml_vec_dot_i2_i8_s` have access to its scale.

Anyway, at least this gives some ideas to try eventually.

Reply:
Got your idea. Assume we have 128 continuous value, we can compact it into (128 / 4) * 7 = 224bits, and other 32bits will be a float32 scale, and still it's 2 bpw. The block size could be 28char(224bits) + 1float32(32bits) == 32bytes. One thing worries me a little is that we need to do some shifting to make the weight align so that we can index from the lookuptable, it may slow down the kernel, but it deserves give try.

Reply:
2024-06-16

> The block size could be 28char(224bits) + 1float32(32bits) == 32bytes. One thing worries me a little is that we need to do some shifting to make the weight align so that we can index from the lookuptable, it may slow down the kernel

Note that the unit for block sizes are elements, while type sizes are in bytes.

To keep the alignment, my suggestion was actually to keep using 8 bits per 4 elements (so that alignment remains easy), but also use the top bit of the first 16 or 32 bytes to store the scale.

To go for maximum compactness, the same idea can be applied to 5-elements per bytes to achieve `1.625 bpw`. The type size would be 13 bytes, the block size 64 elements. 12 bytes of 5 elements per byte and 1 byte of 4 elements, plus part of the scale. But this is slightly more complicated, because 5 isn't a power of 2, so the SIMD vec_dot would need lots of midification, unlike with the other 2 bpw suggestion.